# <p style='text-align: center;'> Segmenting and Clustering Neighborhoods in Toronto </p> 

## Question 1 Creation of the data frame

Importing the libraries

In [1]:
import pandas as pd 
import requests 
!conda install -c conda lxml

Solving environment: ...working... done

# All requested packages already installed.



In [2]:
url = 'https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M'
wiki_url = requests.get(url)
wiki_url

<Response [200]>

Connection established

In [3]:
wiki_data = pd.read_html(wiki_url.text)
wiki_data

[    Postal Code           Borough  \
 0           M1A      Not assigned   
 1           M2A      Not assigned   
 2           M3A        North York   
 3           M4A        North York   
 4           M5A  Downtown Toronto   
 ..          ...               ...   
 175         M5Z      Not assigned   
 176         M6Z      Not assigned   
 177         M7Z      Not assigned   
 178         M8Z         Etobicoke   
 179         M9Z      Not assigned   
 
                                          Neighbourhood  
 0                                         Not assigned  
 1                                         Not assigned  
 2                                            Parkwoods  
 3                                     Victoria Village  
 4                            Regent Park, Harbourfront  
 ..                                                 ...  
 175                                       Not assigned  
 176                                       Not assigned  
 177                

In [4]:
len(wiki_data), type(wiki_data)

(3, list)

Taking the table alone from the webpage

In [5]:
wiki_data = wiki_data[0]
wiki_data

Postal Code           Borough  \
0           M1A      Not assigned   
1           M2A      Not assigned   
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
..          ...               ...   
175         M5Z      Not assigned   
176         M6Z      Not assigned   
177         M7Z      Not assigned   
178         M8Z         Etobicoke   
179         M9Z      Not assigned   

                                         Neighbourhood  
0                                         Not assigned  
1                                         Not assigned  
2                                            Parkwoods  
3                                     Victoria Village  
4                            Regent Park, Harbourfront  
..                                                 ...  
175                                       Not assigned  
176                                       Not assigned  
177                                       Not assigned  
178  Mimico NW, The Queensway West, South of Bloor,...  
179                                       Not assigned  

[180 rows x 3 columns]

Removing Not assigned values from Borough

In [6]:
df = wiki_data[wiki_data["Borough"] != 'Not assigned']
df

Postal Code           Borough  \
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
5           M6A        North York   
6           M7A  Downtown Toronto   
..          ...               ...   
160         M8X         Etobicoke   
165         M4Y  Downtown Toronto   
168         M7Y      East Toronto   
169         M8Y         Etobicoke   
178         M8Z         Etobicoke   

                                         Neighbourhood  
2                                            Parkwoods  
3                                     Victoria Village  
4                            Regent Park, Harbourfront  
5                     Lawrence Manor, Lawrence Heights  
6          Queen's Park, Ontario Provincial Government  
..                                                 ...  
160      The Kingsway, Montgomery Road, Old Mill North  
165                               Church and Wellesley  
168  Business reply mail Processing Centre, South C...  
169  Old Mill South, King's Mill Park, Sunnylea, Hu...  
178  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 3 columns]

Grouping the data by the postal code

In [7]:
df = df.groupby(['Postal Code']).head()
df

Postal Code           Borough  \
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
5           M6A        North York   
6           M7A  Downtown Toronto   
..          ...               ...   
160         M8X         Etobicoke   
165         M4Y  Downtown Toronto   
168         M7Y      East Toronto   
169         M8Y         Etobicoke   
178         M8Z         Etobicoke   

                                         Neighbourhood  
2                                            Parkwoods  
3                                     Victoria Village  
4                            Regent Park, Harbourfront  
5                     Lawrence Manor, Lawrence Heights  
6          Queen's Park, Ontario Provincial Government  
..                                                 ...  
160      The Kingsway, Montgomery Road, Old Mill North  
165                               Church and Wellesley  
168  Business reply mail Processing Centre, South C...  
169  Old Mill South, King's Mill Park, Sunnylea, Hu...  
178  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 3 columns]

Checking for Not assigned values in Neighbourhood column

In [8]:
df.Neighbourhood.str.count("Not assigned").sum()

0

In [9]:
df = df.reset_index()
df

index Postal Code           Borough  \
0        2         M3A        North York   
1        3         M4A        North York   
2        4         M5A  Downtown Toronto   
3        5         M6A        North York   
4        6         M7A  Downtown Toronto   
..     ...         ...               ...   
98     160         M8X         Etobicoke   
99     165         M4Y  Downtown Toronto   
100    168         M7Y      East Toronto   
101    169         M8Y         Etobicoke   
102    178         M8Z         Etobicoke   

                                         Neighbourhood  
0                                            Parkwoods  
1                                     Victoria Village  
2                            Regent Park, Harbourfront  
3                     Lawrence Manor, Lawrence Heights  
4          Queen's Park, Ontario Provincial Government  
..                                                 ...  
98       The Kingsway, Montgomery Road, Old Mill North  
99                                Church and Wellesley  
100  Business reply mail Processing Centre, South C...  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  
102  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 4 columns]

In [10]:
df.drop(['index'], axis = 'columns', inplace = True)
df

Postal Code           Borough  \
0           M3A        North York   
1           M4A        North York   
2           M5A  Downtown Toronto   
3           M6A        North York   
4           M7A  Downtown Toronto   
..          ...               ...   
98          M8X         Etobicoke   
99          M4Y  Downtown Toronto   
100         M7Y      East Toronto   
101         M8Y         Etobicoke   
102         M8Z         Etobicoke   

                                         Neighbourhood  
0                                            Parkwoods  
1                                     Victoria Village  
2                            Regent Park, Harbourfront  
3                     Lawrence Manor, Lawrence Heights  
4          Queen's Park, Ontario Provincial Government  
..                                                 ...  
98       The Kingsway, Montgomery Road, Old Mill North  
99                                Church and Wellesley  
100  Business reply mail Processing Centre, South C...  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  
102  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 3 columns]

Shape of the dataframe

In [11]:
df.shape

(103, 3)

## Question 2 Adding location data

Installing geocoder

In [12]:
pip install geocoder

Note: you may need to restart the kernel to use updated packages.


In [13]:
import geocoder # import geocoder

I was not able to get the geographical coordinates using this code

Using the CSV file now to obtain the coordinates

In [14]:
coor = pd.read_csv("https://cocl.us/Geospatial_data")
coor

Postal Code   Latitude  Longitude
0           M1B  43.806686 -79.194353
1           M1C  43.784535 -79.160497
2           M1E  43.763573 -79.188711
3           M1G  43.770992 -79.216917
4           M1H  43.773136 -79.239476
..          ...        ...        ...
98          M9N  43.706876 -79.518188
99          M9P  43.696319 -79.532242
100         M9R  43.688905 -79.554724
101         M9V  43.739416 -79.588437
102         M9W  43.706748 -79.594054

[103 rows x 3 columns]

In [15]:
print('Shape of dataframe wiki data is', df.shape)
print('Shape of dataframe wiki data is', coor.shape)

Shape of dataframe wiki data is (103, 3)
Shape of dataframe wiki data is (103, 3)


In [16]:
coor.dtypes

Postal Code     object
Latitude       float64
Longitude      float64
dtype: object

In [17]:
df.dtypes

Postal Code      object
Borough          object
Neighbourhood    object
dtype: object

We checked the dtypes and shapes to identify if the dataframes are of the same dimensions so that we can join them

In [18]:
data = df.join(coor.set_index('Postal Code'), on='Postal Code', how = 'inner')
data

Postal Code           Borough  \
0           M3A        North York   
1           M4A        North York   
2           M5A  Downtown Toronto   
3           M6A        North York   
4           M7A  Downtown Toronto   
..          ...               ...   
98          M8X         Etobicoke   
99          M4Y  Downtown Toronto   
100         M7Y      East Toronto   
101         M8Y         Etobicoke   
102         M8Z         Etobicoke   

                                         Neighbourhood   Latitude  Longitude  
0                                            Parkwoods  43.753259 -79.329656  
1                                     Victoria Village  43.725882 -79.315572  
2                            Regent Park, Harbourfront  43.654260 -79.360636  
3                     Lawrence Manor, Lawrence Heights  43.718518 -79.464763  
4          Queen's Park, Ontario Provincial Government  43.662301 -79.389494  
..                                                 ...        ...        ...  
98       The Kingsway, Montgomery Road, Old Mill North  43.653654 -79.506944  
99                                Church and Wellesley  43.665860 -79.383160  
100  Business reply mail Processing Centre, South C...  43.662744 -79.321558  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  43.636258 -79.498509  
102  Mimico NW, The Queensway West, South of Bloor,...  43.628841 -79.520999  

[103 rows x 5 columns]

In [19]:
data.shape

(103, 5)

The combined dataframe is obtained

## Question 3 

Exploring and clustering the neighborhoods in Toronto. Using the New York cluster we obtain inspiration, similar to that
we use KMeans and Foursquare API

In [20]:
!conda install -c conda-forge geopy --yes
from geopy.geocoders import Nominatim

Solving environment: ...working... done

# All requested packages already installed.



In [21]:
address = 'Toronto, Ontario'

geolocator = Nominatim(user_agent="toronto_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of New York City are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of New York City are 43.6534817, -79.3839347.


Latitude and Longitude of Toronto obtained

In [22]:
import folium

In [23]:
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighborhood in zip(data['Latitude'], data['Longitude'], data['Borough'], data['Neighbourhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

Map of Toronto with markers

Initializing FourSquare API credentials

In [24]:
CLIENT_ID = 'CWSCXUF1DOJL1BCTXTIM4LSMTZBQHDMXK2HTYOWQ0YXYNNTH' # your Foursquare ID
CLIENT_SECRET = 'BK4KX4153GRSTRQJIL24WMQZAOOSCI0KADA2EL5LRDM4LVO5' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: CWSCXUF1DOJL1BCTXTIM4LSMTZBQHDMXK2HTYOWQ0YXYNNTH
CLIENT_SECRET:BK4KX4153GRSTRQJIL24WMQZAOOSCI0KADA2EL5LRDM4LVO5


Creating a function to get all venue categories in Toronto

In [25]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius
            )
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighbourhood', 
                  'Neighbourhood Latitude', 
                  'Neighbourhood Longitude', 
                  'Venue', 
                  'Venue Category']
    
    return(nearby_venues)

Collecting venues in Toronto in each neighbourhood

In [26]:
venues_in_toronto = getNearbyVenues(data['Neighbourhood'], data['Latitude'], data['Longitude'])

Parkwoods
Victoria Village
Regent Park, Harbourfront
Lawrence Manor, Lawrence Heights
Queen's Park, Ontario Provincial Government
Islington Avenue, Humber Valley Village
Malvern, Rouge
Don Mills
Parkview Hill, Woodbine Gardens
Garden District, Ryerson
Glencairn
West Deane Park, Princess Gardens, Martin Grove, Islington, Cloverdale
Rouge Hill, Port Union, Highland Creek
Don Mills
Woodbine Heights
St. James Town
Humewood-Cedarvale
Eringate, Bloordale Gardens, Old Burnhamthorpe, Markland Wood
Guildwood, Morningside, West Hill
The Beaches
Berczy Park
Caledonia-Fairbanks
Woburn
Leaside
Central Bay Street
Christie
Cedarbrae
Hillcrest Village
Bathurst Manor, Wilson Heights, Downsview North
Thorncliffe Park
Richmond, Adelaide, King
Dufferin, Dovercourt Village
Scarborough Village
Fairview, Henry Farm, Oriole
Northwood Park, York University
East Toronto, Broadview North (Old East York)
Harbourfront East, Union Station, Toronto Islands
Little Portugal, Trinity
Kennedy Park, Ionview, East Birchmo

In [27]:
venues_in_toronto.shape

(1334, 5)

Checking the shape, thus we find we have 1334 venues

In [28]:
venues_in_toronto.head()

Neighbourhood  Neighbourhood Latitude  Neighbourhood Longitude  \
0         Parkwoods               43.753259               -79.329656   
1         Parkwoods               43.753259               -79.329656   
2  Victoria Village               43.725882               -79.315572   
3  Victoria Village               43.725882               -79.315572   
4  Victoria Village               43.725882               -79.315572   

                    Venue         Venue Category  
0         Brookbanks Park                   Park  
1           Variety Store      Food & Drink Shop  
2  Victoria Village Arena           Hockey Arena  
3             Tim Hortons            Coffee Shop  
4               Portugril  Portuguese Restaurant

Checking a sample of the data

In [29]:
venues_in_toronto.groupby('Neighbourhood').head()

Neighbourhood  \
0                                             Parkwoods   
1                                             Parkwoods   
2                                      Victoria Village   
3                                      Victoria Village   
4                                      Victoria Village   
...                                                 ...   
1318  Mimico NW, The Queensway West, South of Bloor,...   
1319  Mimico NW, The Queensway West, South of Bloor,...   
1320  Mimico NW, The Queensway West, South of Bloor,...   
1321  Mimico NW, The Queensway West, South of Bloor,...   
1322  Mimico NW, The Queensway West, South of Bloor,...   

      Neighbourhood Latitude  Neighbourhood Longitude                   Venue  \
0                  43.753259               -79.329656         Brookbanks Park   
1                  43.753259               -79.329656           Variety Store   
2                  43.725882               -79.315572  Victoria Village Arena   
3                  43.725882               -79.315572             Tim Hortons   
4                  43.725882               -79.315572               Portugril   
...                      ...                      ...                     ...   
1318               43.628841               -79.520999              Wingporium   
1319               43.628841               -79.520999        South St. Burger   
1320               43.628841               -79.520999               Dollarama   
1321               43.628841               -79.520999          Healthy Planet   
1322               43.628841               -79.520999    Artisano Bakery Café   

             Venue Category  
0                      Park  
1         Food & Drink Shop  
2              Hockey Arena  
3               Coffee Shop  
4     Portuguese Restaurant  
...                     ...  
1318            Wings Joint  
1319           Burger Joint  
1320         Discount Store  
1321        Supplement Shop  
1322                 Bakery  

[406 rows x 5 columns]

Grouping by neighbourhoods

In [30]:
venues_in_toronto.groupby('Venue Category').max()

Neighbourhood  \
Venue Category                                                          
Accessories Store                    Lawrence Manor, Lawrence Heights   
Airport                                                     Downsview   
Airport Food Court  CN Tower, King and Spadina, Railway Lands, Har...   
Airport Gate        CN Tower, King and Spadina, Railway Lands, Har...   
Airport Lounge      CN Tower, King and Spadina, Railway Lands, Har...   
...                                                               ...   
Warehouse Store                                      Thorncliffe Park   
Wine Bar                                              Studio District   
Wings Joint         Mimico NW, The Queensway West, South of Bloor,...   
Women's Store                                     Caledonia-Fairbanks   
Yoga Studio                            University of Toronto, Harbord   

                    Neighbourhood Latitude  Neighbourhood Longitude  \
Venue Category                                                        
Accessories Store                43.718518               -79.464763   
Airport                          43.737473               -79.394420   
Airport Food Court               43.628947               -79.394420   
Airport Gate                     43.628947               -79.394420   
Airport Lounge                   43.628947               -79.394420   
...                                    ...                      ...   
Warehouse Store                  43.705369               -79.349372   
Wine Bar                         43.659526               -79.340923   
Wings Joint                      43.628841               -79.520999   
Women's Store                    43.689026               -79.453512   
Yoga Studio                      43.715383               -79.321558   

                                              Venue  
Venue Category                                       
Accessories Store               Ardene Shoes Outlet  
Airport             Toronto Downsview Airport (YZD)  
Airport Food Court                Billy Bishop Café  
Airport Gate                                 Gate 8  
Airport Lounge                        Porter Lounge  
...                                             ...  
Warehouse Store                              Costco  
Wine Bar                            Paris Paris Bar  
Wings Joint                              Wingporium  
Women's Store                         Maximum Woman  
Yoga Studio                               YogaSpace  

[234 rows x 4 columns]

Max number of venues grouped by its venue category

### One Hot Encoding

In [31]:
toronto_onehot = pd.get_dummies(venues_in_toronto[['Venue Category']], prefix="", prefix_sep="")
toronto_onehot

Accessories Store  Airport  Airport Food Court  Airport Gate  \
0                     0        0                   0             0   
1                     0        0                   0             0   
2                     0        0                   0             0   
3                     0        0                   0             0   
4                     0        0                   0             0   
...                 ...      ...                 ...           ...   
1329                  0        0                   0             0   
1330                  0        0                   0             0   
1331                  0        0                   0             0   
1332                  0        0                   0             0   
1333                  0        0                   0             0   

      Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0                  0                0                 0                    0   
1                  0                0                 0                    0   
2                  0                0                 0                    0   
3                  0                0                 0                    0   
4                  0                0                 0                    0   
...              ...              ...               ...                  ...   
1329               0                0                 0                    0   
1330               0                0                 0                    0   
1331               0                0                 0                    0   
1332               0                0                 0                    0   
1333               0                0                 0                    0   

      Antique Shop  Aquarium  ...  Train Station  \
0                0         0  ...              0   
1                0         0  ...              0   
2                0         0  ...              0   
3                0         0  ...              0   
4                0         0  ...              0   
...            ...       ...  ...            ...   
1329             0         0  ...              0   
1330             0         0  ...              0   
1331             0         0  ...              0   
1332             0         0  ...              0   
1333             0         0  ...              0   

      Vegetarian / Vegan Restaurant  Video Game Store  Video Store  \
0                                 0                 0            0   
1                                 0                 0            0   
2                                 0                 0            0   
3                                 0                 0            0   
4                                 0                 0            0   
...                             ...               ...          ...   
1329                              0                 0            0   
1330                              0                 0            0   
1331                              0                 0            0   
1332                              0                 0            0   
1333                              0                 0            0   

      Vietnamese Restaurant  Warehouse Store  Wine Bar  Wings Joint  \
0                         0                0         0            0   
1                         0                0         0            0   
2                         0                0         0            0   
3                         0                0         0            0   
4                         0                0         0            0   
...                     ...              ...       ...          ...   
1329                      0                0         0            0   
1330                      0                0         0            0   
1331                      0                0         0            0   
1332                      0 

In [32]:
toronto_onehot.shape

(1334, 234)

Adding neighbourhood to one hot 

In [33]:
toronto_onehot['Neighbourhood'] = venues_in_toronto['Neighbourhood'] 

# moving neighborhood column to the first column
fixed_columns = [toronto_onehot.columns[-1]] + list(toronto_onehot.columns[:-1])
toronto_onehot = toronto_onehot[fixed_columns]

toronto_onehot.head()

Neighbourhood  Accessories Store  Airport  Airport Food Court  \
0         Parkwoods                  0        0                   0   
1         Parkwoods                  0        0                   0   
2  Victoria Village                  0        0                   0   
3  Victoria Village                  0        0                   0   
4  Victoria Village                  0        0                   0   

   Airport Gate  Airport Lounge  Airport Service  Airport Terminal  \
0             0               0                0                 0   
1             0               0                0                 0   
2             0               0                0                 0   
3             0               0                0                 0   
4             0               0                0                 0   

   American Restaurant  Antique Shop  ...  Train Station  \
0                    0             0  ...              0   
1                    0             0  ...              0   
2                    0             0  ...              0   
3                    0             0  ...              0   
4                    0             0  ...              0   

   Vegetarian / Vegan Restaurant  Video Game Store  Video Store  \
0                              0                 0            0   
1                              0                 0            0   
2                              0                 0            0   
3                              0                 0            0   
4                              0                 0            0   

   Vietnamese Restaurant  Warehouse Store  Wine Bar  Wings Joint  \
0                      0                0         0            0   
1                      0                0         0            0   
2                      0                0         0            0   
3                      0                0         0            0   
4                      0                0         0            0   

   Women's Store  Yoga Studio  
0              0            0  
1              0            0  
2              0            0  
3              0            0  
4              0            0  

[5 rows x 235 columns]

Grouping by Neighbourhood

In [34]:
toronto_grouped = toronto_onehot.groupby('Neighbourhood').mean().reset_index()
toronto_grouped.head()

Neighbourhood  Accessories Store  \
0                                        Agincourt                0.0   
1                           Alderwood, Long Branch                0.0   
2  Bathurst Manor, Wilson Heights, Downsview North                0.0   
3                                  Bayview Village                0.0   
4                Bedford Park, Lawrence Manor East                0.0   

   Airport  Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0      0.0                 0.0           0.0             0.0              0.0   
1      0.0                 0.0           0.0             0.0              0.0   
2      0.0                 0.0           0.0             0.0              0.0   
3      0.0                 0.0           0.0             0.0              0.0   
4      0.0                 0.0           0.0             0.0              0.0   

   Airport Terminal  American Restaurant  Antique Shop  ...  Train Station  \
0               0.0                 0.00           0.0  ...            0.0   
1               0.0                 0.00           0.0  ...            0.0   
2               0.0                 0.00           0.0  ...            0.0   
3               0.0                 0.00           0.0  ...            0.0   
4               0.0                 0.04           0.0  ...            0.0   

   Vegetarian / Vegan Restaurant  Video Game Store  Video Store  \
0                            0.0               0.0          0.0   
1                            0.0               0.0          0.0   
2                            0.0               0.0          0.0   
3                            0.0               0.0          0.0   
4                            0.0               0.0          0.0   

   Vietnamese Restaurant  Warehouse Store  Wine Bar  Wings Joint  \
0                    0.0              0.0       0.0          0.0   
1                    0.0              0.0       0.0          0.0   
2                    0.0              0.0       0.0          0.0   
3                    0.0              0.0       0.0          0.0   
4                    0.0              0.0       0.0          0.0   

   Women's Store  Yoga Studio  
0            0.0          0.0  
1            0.0          0.0  
2            0.0          0.0  
3            0.0          0.0  
4            0.0          0.0  

[5 rows x 235 columns]

A function to return most common venues

In [35]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [36]:
import numpy as np

Taking top 10 cluster the neighborhoods of venue categories

In [42]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighbourhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighbourhood'] = toronto_grouped['Neighbourhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighbourhood 1st Most Common Venue  \
0                                        Agincourt                Lounge   
1                           Alderwood, Long Branch           Pizza Place   
2  Bathurst Manor, Wilson Heights, Downsview North           Coffee Shop   
3                                  Bayview Village                  Café   
4                Bedford Park, Lawrence Manor East           Coffee Shop   

       2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0  Latin American Restaurant          Skating Rink        Breakfast Spot   
1                Coffee Shop                   Pub                   Gym   
2                       Bank         Grocery Store     Mobile Phone Shop   
3                       Bank    Chinese Restaurant   Japanese Restaurant   
4                Pizza Place        Sandwich Place    Italian Restaurant   

  5th Most Common Venue        6th Most Common Venue  \
0           Yoga Studio  Eastern European Restaurant   
1        Sandwich Place                 Skating Rink   
2         Shopping Mall                  Bridal Shop   
3      Department Store            Electronics Store   
4       Thai Restaurant                Grocery Store   

         7th Most Common Venue    8th Most Common Venue 9th Most Common Venue  \
0          Dumpling Restaurant                Drugstore            Donut Shop   
1             Department Store             Dessert Shop    Dim Sum Restaurant   
2               Sandwich Place            Deli / Bodega            Restaurant   
3  Eastern European Restaurant      Dumpling Restaurant             Drugstore   
4             Sushi Restaurant  Comfort Food Restaurant             Pet Store   

  10th Most Common Venue  
0                Dog Run  
1                  Diner  
2            Supermarket  
3             Donut Shop  
4               Pharmacy

Lets make a model to cluster 

In [43]:
# import k-means from clustering stage
from sklearn.cluster import KMeans

In [44]:
# set number of clusters
k_num_clusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighbourhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=k_num_clusters, random_state=0).fit(toronto_grouped_clustering)
kmeans

KMeans(algorithm='auto', copy_x=True, init='k-means++', max_iter=300,
       n_clusters=5, n_init=10, n_jobs=None, precompute_distances='auto',
       random_state=0, tol=0.0001, verbose=0)

Checking labels

In [45]:
kmeans.labels_[0:100]

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 2, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 2, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 3, 0, 0, 0, 0,
       2, 0, 2, 0, 0, 1, 2, 0, 0, 0, 0, 0, 0, 0, 3, 0, 0, 2, 0, 0, 0, 2,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 4, 0,
       0, 2, 0, 0, 0, 0, 2, 2])

In [46]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

toronto_merged = data

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighbourhood'), on='Neighbourhood')

toronto_merged.head() # check the last columns!

Postal Code           Borough                                Neighbourhood  \
0         M3A        North York                                    Parkwoods   
1         M4A        North York                             Victoria Village   
2         M5A  Downtown Toronto                    Regent Park, Harbourfront   
3         M6A        North York             Lawrence Manor, Lawrence Heights   
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   

    Latitude  Longitude  Cluster Labels 1st Most Common Venue  \
0  43.753259 -79.329656             2.0                  Park   
1  43.725882 -79.315572             0.0     French Restaurant   
2  43.654260 -79.360636             0.0           Coffee Shop   
3  43.718518 -79.464763             0.0        Clothing Store   
4  43.662301 -79.389494             0.0           Coffee Shop   

    2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0       Food & Drink Shop           Yoga Studio         Deli / Bodega   
1             Coffee Shop          Intersection          Hockey Arena   
2                    Park        Breakfast Spot               Theater   
3  Furniture / Home Store     Accessories Store           Event Space   
4             Yoga Studio    College Auditorium              Beer Bar   

         5th Most Common Venue       6th Most Common Venue  \
0  Eastern European Restaurant         Dumpling Restaurant   
1        Portuguese Restaurant  Financial or Legal Service   
2                         Café                      Bakery   
3                  Coffee Shop       Vietnamese Restaurant   
4                Smoothie Shop              Sandwich Place   

  7th Most Common Venue  8th Most Common Venue 9th Most Common Venue  \
0             Drugstore             Donut Shop               Dog Run   
1           Pizza Place                Dog Run   Distribution Center   
2   Distribution Center                    Spa          Dessert Shop   
3              Boutique    Distribution Center    Dim Sum Restaurant   
4                  Café  Portuguese Restaurant    Chinese Restaurant   

  10th Most Common Venue  
0    Distribution Center  
1             Donut Shop  
2         Farmers Market  
3                  Diner  
4                   Park

In [57]:
toronto_merged_nonan = toronto_merged.dropna(subset=['Cluster Labels'])

In [58]:
import matplotlib.cm as cm
import matplotlib.colors as colors

In [60]:
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(k_num_clusters)
ys = [i + x + (i*x)**2 for i in range(k_num_clusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged_nonan['Latitude'], toronto_merged_nonan['Longitude'], toronto_merged_nonan['Neighbourhood'], toronto_merged_nonan['Cluster Labels']):
    label = folium.Popup('Cluster ' + str(int(cluster) +1) + '\n' + str(poi) , parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[int(cluster-1)],
        fill=True,
        fill_color=rainbow[int(cluster-1)]
        ).add_to(map_clusters)
        
map_clusters

The clusters are plotted on the map

Lets verify each cluster

Cluster 1 

In [61]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 0, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]


Borough  Cluster Labels 1st Most Common Venue  \
1          North York             0.0     French Restaurant   
2    Downtown Toronto             0.0           Coffee Shop   
3          North York             0.0        Clothing Store   
4    Downtown Toronto             0.0           Coffee Shop   
7          North York             0.0                   Gym   
..                ...             ...                   ...   
97   Downtown Toronto             0.0                  Café   
98          Etobicoke             0.0                 River   
99   Downtown Toronto             0.0           Coffee Shop   
100      East Toronto             0.0    Light Rail Station   
102         Etobicoke             0.0         Grocery Store   

      2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
1               Coffee Shop          Intersection          Hockey Arena   
2                      Park        Breakfast Spot               Theater   
3    Furniture / Home Store     Accessories Store           Event Space   
4               Yoga Studio    College Auditorium              Beer Bar   
7               Coffee Shop            Beer Store   Sporting Goods Shop   
..                      ...                   ...                   ...   
97              Coffee Shop            Restaurant    Seafood Restaurant   
98                     Park            Smoke Shop                  Pool   
99                     Park            Steakhouse             Bookstore   
100             Yoga Studio           Pizza Place      Recording Studio   
102  Thrift / Vintage Store     Convenience Store         Burrito Place   

     5th Most Common Venue       6th Most Common Venue 7th Most Common Venue  \
1    Portuguese Restaurant  Financial or Legal Service           Pizza Place   
2                     Café                      Bakery   Distribution Center   
3              Coffee Shop       Vietnamese Restaurant              Boutique   
4            Smoothie Shop              Sandwich Place                  Café   
7      Japanese Restaurant                  Restaurant  Caribbean Restaurant   
..                     ...                         ...                   ...   
97     Japanese Restaurant                         Pub             Speakeasy   
98             Curling Ice                   Drugstore            Donut Shop   
99          Breakfast Spot          Salon / Barbershop       Bubble Tea Shop   
100             Skate Park               Auto Workshop  Fast Food Restaurant   
102           Burger Joint              Discount Store        Sandwich Place   

     8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
1                  Dog Run   Distribution Center             Donut Shop  
2                      Spa          Dessert Shop         Farmers Market  
3      Distribution Center    Dim Sum Restaurant                  Diner  
4    Portuguese Restaurant    Chinese Restaurant                   Park  
7           Sandwich Place       Bubble Tea Shop              Bike Shop  
..                     ...                   ...                    ...  
97              Steakhouse                   Bar                  Hotel  
98                 Dog Run   Distribution Center         Discount Store  
99              Restaurant      Ramen Restaurant                    Pub  
100         Farmers Market         Burrito Place                    Spa  
102             Kids Store  Fast Food Restaurant                 Bakery  

[86 rows x 12 columns]

Cluster 2

In [64]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 1, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
6  Scarborough             1.0  Fast Food Restaurant      Department Store   

  3rd Most Common Venue 4th Most Common Venue        5th Most Common Venue  \
6           Event Space     Electronics Store  Eastern European Restaurant   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
6   Dumpling Restaurant             Drugstore            Donut Shop   

  9th Most Common Venue 10th Most Common Venue  
6               Dog Run    Distribution Center

Cluster 3

In [63]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 2, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]


Borough  Cluster Labels 1st Most Common Venue  \
0         North York             2.0                  Park   
21              York             2.0                  Park   
35         East York             2.0                  Park   
45        North York             2.0                  Park   
52        North York             2.0                  Park   
61   Central Toronto             2.0                  Park   
66        North York             2.0                  Park   
77         Etobicoke             2.0     Mobile Phone Shop   
85       Scarborough             2.0            Playground   
91  Downtown Toronto             2.0                  Park   

         2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0            Food & Drink Shop           Yoga Studio         Deli / Bodega   
21               Women's Store         Deli / Bodega     Electronics Store   
35           Convenience Store          Intersection           Yoga Studio   
45         Martial Arts School           Yoga Studio         Deli / Bodega   
52                Home Service           Yoga Studio         Deli / Bodega   
61                 Swim School              Bus Line           Yoga Studio   
66  Construction & Landscaping     Convenience Store           Yoga Studio   
77                        Park        Sandwich Place           Yoga Studio   
85                        Park           Yoga Studio         Deli / Bodega   
91                  Playground                 Trail           Yoga Studio   

          5th Most Common Venue        6th Most Common Venue  \
0   Eastern European Restaurant          Dumpling Restaurant   
21  Eastern European Restaurant          Dumpling Restaurant   
35             Department Store            Electronics Store   
45  Eastern European Restaurant          Dumpling Restaurant   
52            Electronics Store  Eastern European Restaurant   
61             Department Store  Eastern European Restaurant   
66             Department Store            Electronics Store   
77          Dumpling Restaurant                    Drugstore   
85  Eastern European Restaurant          Dumpling Restaurant   
91                 Dance Studio          Dumpling Restaurant   

          7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
0                     Drugstore            Donut Shop               Dog Run   
21                    Drugstore            Donut Shop               Dog Run   
35  Eastern European Restaurant   Dumpling Restaurant             Drugstore   
45                    Drugstore            Donut Shop               Dog Run   
52          Dumpling Restaurant             Drugstore            Donut Shop   
61          Dumpling Restaurant             Drugstore            Donut Shop   
66  Eastern European Restaurant   Dumpling Restaurant             Drugstore   
77                   Donut Shop               Dog Run   Distribution Center   
85                    Drugstore            Donut Shop               Dog Run   
91                    Drugstore            Donut Shop               Dog Run   

   10th Most Common Venue  
0     Distribution Center  
21    Distribution Center  
35             Donut Shop  
45    Distribution Center  
52                Dog Run  
61                Dog Run  
66             Donut Shop  
77         Discount Store  
85    Distribution Center  
91    Distribution Center

Cluster 4

In [65]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 3, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
57   North York             3.0        Baseball Field           Yoga Studio   
101   Etobicoke             3.0        Baseball Field           Yoga Studio   

    3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
57         Farmers Market           Event Space     Electronics Store   
101        Farmers Market           Event Space     Electronics Store   

           6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
57   Eastern European Restaurant   Dumpling Restaurant             Drugstore   
101  Eastern European Restaurant   Dumpling Restaurant             Drugstore   

    9th Most Common Venue 10th Most Common Venue  
57             Donut Shop                Dog Run  
101            Donut Shop                Dog Run

Cluster 5

In [66]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 4, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
11  Etobicoke             4.0           Golf Course           Yoga Studio   

   3rd Most Common Venue 4th Most Common Venue        5th Most Common Venue  \
11         Deli / Bodega     Electronics Store  Eastern European Restaurant   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
11   Dumpling Restaurant             Drugstore            Donut Shop   

   9th Most Common Venue 10th Most Common Venue  
11               Dog Run    Distribution Center

All clusters have been verified